server/internal/server/ollama_client.go

```diff
@@ -1,72 +1,133 @@
+			Scheme: "http",
+			Host:   os.Getenv("OLLAMA_HOST"),
+		},
+		http: http.DefaultClient,
 	}
-	return client, nil
 }
 
-func handleChatRequest(ctx context.Context, req *v1.CreateChatCompletionRequest) (*v1.ChatCompletion, error) {
-	client, err := newClient()
+const maxBufferSize = 512 * 1000
+
+func (c *client) stream(ctx context.Context, method, path string, data any, fn func([]byte) error) error {
+	var buf *bytes.Buffer
+	if data != nil {
+		bts, err := json.Marshal(data)
+		if err != nil {
+			return err
+		}
+
+		buf = bytes.NewBuffer(bts)
+	}
+
+	requestURL := c.base.JoinPath(path)
+	request, err := http.NewRequestWithContext(ctx, method, requestURL.String(), buf)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Failed to create a client: %s", err)
+		return err
 	}
 
-	var msgs []api.Message
-	for _, msg := range req.Messages {
-		msgs = append(msgs, api.Message{
-			Role:    msg.Role,
-			Content: msg.Content,
-		})
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Set("Accept", "application/x-ndjson")
+
+	response, err := c.http.Do(request)
+	if err != nil {
+		return err
 	}
+	defer response.Body.Close()
+
+	scanner := bufio.NewScanner(response.Body)
+	// increase the buffer size to avoid running out of space
+	scanBuf := make([]byte, 0, maxBufferSize)
+	scanner.Buffer(scanBuf, maxBufferSize)
+	for scanner.Scan() {
+		var errorResponse struct {
+			Error string `json:"error,omitempty"`
+		}
+
+		bts := scanner.Bytes()
+		if err := json.Unmarshal(bts, &errorResponse); err != nil {
+			return fmt.Errorf("unmarshal: %w", err)
+		}
 
-	ollamaReq := &api.ChatRequest{
-		Model:    req.Model,
-		Messages: msgs,
+		if errorResponse.Error != "" {
+			return fmt.Errorf(errorResponse.Error)
+		}
+
+		if response.StatusCode >= http.StatusBadRequest {
+			return status.Errorf(codes.Code(response.StatusCode), "Status: %s: %s", response.Status, errorResponse.Error)
+		}
+
+		if err := fn(bts); err != nil {
+			return err
+		}
 	}
 
-	var ollamaResp api.ChatResponse
-	fn := func(resp api.ChatResponse) error {
-		ollamaResp = resp
-		return processChatResponse(resp)
+	return nil
+}
+
+// chatResponseFunc is a function that [client.Chat] invokes every time
+// a response is received from the service. If this function returns an error,
+// [client.Chat] will stop generating and return this error.
+type chatResponseFunc func(*v1.ChatCompletion) error
+
+// Chat generates the next message in a chat.
+func (c *client) Chat(ctx context.Context, req *v1.CreateChatCompletionRequest, fn chatResponseFunc) error {
+	return c.stream(ctx, http.MethodPost, "/v1/chat/completions", req, func(bts []byte) error {
+		var resp v1.ChatCompletion
+		if err := json.Unmarshal(bts, &resp); err != nil {
+			return err
+		}
+
+		return fn(&resp)
+	})
+}
+
+func handleChatRequest(ctx context.Context, req *v1.CreateChatCompletionRequest) (*v1.ChatCompletion, error) {
+	client := newClient()
+
+	var resp *v1.ChatCompletion
```

Comment:
It seems to me that this works only when `fn` is called only once? I thought `fn` can be called multiple times.

Reply:
not very sure what you mean by "called only once"? Here `processChatResponse` is expected to be a read-only function of `resp`, so no need to create a copy of original `resp`.

Reply:
So line 88 calls `fn()` inside `scanner.Scan()`. Whenever `fn` is called we update `resp` here. Then we return `resp` at the end of this function. 

Is the assumption here that we only need to return the last `resp` as RPC response?

Reply:
yeah i'm not sure if gRPC streaming works as it might not be compatible with the OpenAI API. We might need to just implement HTTP handlers without gRPC.